### PR TITLE
Cannon state versioning

### PR DESCRIPTION
**Pull request by Inphi**
Proposal for handling changes to the offchain Cannon STF.

### CODE REVIEW

File: cannon-state-versioning.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+Once the CLI has identified the appropriate Cannon VM binary for the given prestate, it loads this binary via go:embed.
+The selected Cannon program is then executed using [`execve(2)`](https://man7.org/linux/man-pages/man2/execve.2.html).
+This execution method allows for efficient and isolated running of the chosen Cannon version, ensuring that each prestate is processed by the correct VM implementation.
+
+An additional benefit of this approach is it facilitates multi-architecture MIPS Cannon. Particularly for the usecase where multiple Cannon binaries with different [build tags are needed to compile Cannon for MIPS32 and MIPS64 VMs](https://github.com/ethereum-optimism/optimism/pull/12029).
+
+## State Version
+
+This approach implies that the `StateVersion` must be used to also encode the VM version. As such, every change to the STF, no matter how little will require a new version.
+However, `StateVersion` is a uint8 field and it's not terribly unlikely that there are under 256 VM implementations.
```
Comment (mbaxter) [inline cannon-state-versioning.md:55]:
```suggestion
However, `StateVersion` is a uint8 field and it's not terribly unlikely that there are over 256 VM implementations. 
```

**Review by clabby — APPROVED** [review]
LGTM, nice

**Comment by tynes** [comment]
This is sick

File: cannon-state-versioning.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+## State Version
+
+This approach implies that the `StateVersion` must be used to also encode the VM version. As such, every change to the STF, no matter how little will require a new version.
+However, `StateVersion` is a uint8 field and it's not terribly unlikely that there are over 256 VM implementations. 
+As such, we need a new wider field in the encoded state file to encapsulate the VM version.
+
+## Supporting new cannon embeds 
+
+While every version of cannon can be embedded into the cannon CLI, this can result in a really large binary.
+Instead, the Cannon CLI only needs to maintain the VM implemnetations that it's likely to use.
```
Comment (pegahcarter) [inline cannon-state-versioning.md:61]: nit: implementations

**Comment by pauldowman** [comment]
Is there anything left to do on this doc? It seems ready to for any last reviews and to be merged, especially since the code changes are already merged. 🙂

**Comment by Inphi** [comment]
> Is there anything left to do on this doc? It seems ready to for any last reviews and to be merged, especially since the code changes are already merged. 🙂

it's good to merge. I can do so tomorrow

**Pull request merged**